### PR TITLE
Disallow final value in logic functions and ghost

### PR DESCRIPTION
**Pull request by voidc**
This extends the purity check to consider three levels of purity: logic, ghost, and program.
Specifications (function pre- and postconditions and loop invariants) and predicate functions constitute a context of logic purity and can thus call predicate, logic, and builtin functions, and use pearlite operators.
Ghost blocks and logic functions constitute a context of ghost purity, which permits the same operations as the logic context except using the final value operator.
Lastly, every other function constitutes a context of program purity, in which only other program functions may be called.

### CODE REVIEW

File: creusot/src/translation/pearlite.rs [hunk 1]
```diff
@@ -225,8 +225,9 @@ pub(crate) fn pearlite<'tcx>(
         return Err(Error::new(ctx.def_span(id), "type checking failed"));
     };
 
+    // let purity = Purity::of_def_id(ctx.tcx, id.to_def_id());
     visit::walk_expr(
```
Comment (voidc) [inline creusot/src/translation/pearlite.rs:229]: Why do we have to check purity again here? Shouldn't `before_analysis` already check every HIR item?

**Comment by voidc** [comment]
We also cannot allow calling predicates in ghost context as this would again permit depending on prophecies. However, we commonly call functions like `lt_log` in logic functions, which means changing them from predicates to logic functions. Ideally, such functions would be `pure` functions that can be called in any context.

File: creusot/src/translation/pearlite.rs (hunk 1 above)
Comment (xldenis) [reply creusot/src/translation/pearlite.rs:229]: hmm, i think its a relic of times long past.

**Comment by xldenis** [comment]
This looks good to me! Perhaps we could also rename `logic` to `ghost` (or `ghost_fn` to make it easier for now) in a future PR, and `predicate` to `logic`?

**Comment by xldenis** [comment]
Yea I think the idea would be to rename predicate to logic and lift the restriction on return types.

**Comment by sarsko** [comment]
Hmm, I'm quite a fan of the `logic`/`predicate` distinction. They allow for easier grouping, which I have found valuable as a proof grows, and it is also useful when reading a proof written by others. Tbh, I would even advocate for the addition of `#[lemma]`, to enable further grouping and avoid the current scheme of:
```rust
#[logic]
pub fn lemma_bla_blabla() {}
```

I don't see the case for enabling predicates to return non-booleans. What's the reasoning behind this / why would this be useful?

**Comment by dewert99** [comment]
Basically, the `ghost!` macro is currently unsound since it can call `logic` functions and they can use the final `^` operator (see #436).  One solution to the problem would be to disallow the final `^` operator in `logic` functions and only allow it in predicates, this could potentially lead to cases where someone wants to use `^` in a function and return a non-boolean.

Maybe instead of framing things in terms of renaming, we should just say we're adding a new `ghost` attribute that works like `logic` but disallows the final operator and then only allow `ghost` functions in `ghost!` macros.

**Pull request merged**